operator: Improve identity GC efficiency

christarazi opened the pull request:
With this commit, the identity GC rate limit
(--identity-gc-rate-interval) becomes the effective rate at which
identities are garbage collected. Previously, the identity GC interval
(--identity-gc-interval) would cause the Operator to GC for that much
time, then the sleep for that much time, rinse and repeat, effectively
halving the rate.

To use concrete numbers for an example, let's say our interval is 5m and
our GC rate interval is 1000 per minute. It would mean that previously,
we would GC 5000 identities at a maximum for 10m (assuming that deletion
takes 0s). How was that calculated? Each minute, we GC 1000 identities.
After 5m, we have GC'd 5000 identities. But now we have to sleep for 5m
because that's our GC interval. Hence making our effective GC rate 500
per minute (instead of being 1000/m).

Now, we compute the time taken to perform the actual GC and subtract
that from the interval. So in our above example, we eliminate the dead
time of 5m and avoid slashing our effective GC rate in half. This change
allows the Operator to keep up with the demand more efficiently. The
Operator will warn if the GC duration took longer than the interval and
set the sleep duration to 0.

Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: Dan Wendlandt <dan@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
operator: Improve identity GC efficiency

With this commit, the identity GC rate limit
(--identity-gc-rate-interval) becomes the effective rate at which
identities are garbage collected. Previously, the identity GC interval
(--identity-gc-interval) would cause the Operator to GC for that much
time, then the sleep for that much time, rinse and repeat, effectively
halving the rate.

To use concrete numbers for an example, let's say our interval is 5m and
our GC rate interval is 1000 per minute. It would mean that previously,
we would GC 5000 identities at a maximum for 10m (assuming that deletion
takes 0s). How was that calculated? Each minute, we GC 1000 identities.
After 5m, we have GC'd 5000 identities. But now we have to sleep for 5m
because that's our GC interval. Hence making our effective GC rate 500
per minute (instead of being 1000/m).

Now, we compute the time taken to perform the actual GC and subtract
that from the interval. So in our above example, we eliminate the dead
time of 5m and avoid slashing our effective GC rate in half. This change
allows the Operator to keep up with the demand more efficiently. The
Operator will warn if the GC duration took longer than the interval and
set the sleep duration to 0.

Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: Dan Wendlandt <dan@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on operator/identity_gc.go:
```diff
@@ -50,7 +54,19 @@ func startKvstoreIdentityGC() {
+				log.WithFields(logrus.Fields{
+					logfields.Interval: operatorOption.Config.IdentityGCInterval,
+					logfields.Duration: gcDuration,
+				}).Warning("Garbage collection took longer than the GC interval")
```
I wonder if we can make this clearer, eg
```suggestion
				log.WithFields(logrus.Fields{
					logfields.Interval: operatorOption.Config.IdentityGCInterval,
					logfields.Duration: gcDuration,
					logfields.Hint: "Is there a ratelimit configured on the kvstore client or server?",
				}).Warningf("Identity garbage collection took longer than the %s", operatorOption.IdentityGCInterval)
```

(note `logfields.Hint` is somehow not yet a thing, but maybe it should be).

joestringer replied:
I also thought a little about the log level, and Warning seems right to me. This is not inherently an error per se, and I might even expect that Cilium can continue to operate successfully for some time after issuing this warning. But left unchecked this could escalate to a more serious error (again, not because of this logic, but because of other logic elsewhere that depends on this logic keeping up).

christarazi replied:
Implemented feedback, thanks

joestringer reviewed (changes requested):
Just one minor nit to make the warning a little bit more clear.

Noting also what we found on friday, which was that etcdconfig.etcd.qps can be a limiting factor here.

christarazi committed:
logfields: Add Hint field

This is useful in warning or error level messages to help nudge the user
in the right direction when troubleshooting.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on operator/identity_gc.go:
```diff
@@ -50,7 +54,20 @@ func startKvstoreIdentityGC() {
+			} else {
+				sleep -= gcDuration
+			}
+			<-gcTimer.After(sleep)
```
Nit, we could move this to line 67 and we wouldn't need to run `gcTimer.After` with `0`.

christarazi replied:
Fixed, thanks

aanm reviewed (approved):
LGTM, only a small nit.
